### PR TITLE
refactor version metadata fetching

### PR DESCRIPTION
**Pull request by tomerd**
motivation: move the version metadata fetching closer to download of archive so we can reuse this data to explose publisher information more easily

changes:
* refactor TOFO and signature validation utilities to take a provider of metadata instead of dependending directly on the client
* refactor archive downloading to first fecth the version metadata (cached)
* update call sites and tests

### CODE REVIEW

**Comment by tomerd** [comment]
cc @yim-lee

**Comment by tomerd** [comment]
@swift-ci smoke test

File: Sources/PackageRegistry/SignatureValidation.swift [hunk 1]
```diff
@@ -207,59 +238,56 @@ struct SignatureValidation {
...
-                completion(.failure(RegistryError.failedRetrievingSourceArchiveSignature(
-                    registry: registry,
-                    package: package.underlying,
-                    version: version,
-                    error: actualError
-                )))
-            }
-        }
-    }
+    /*
```
Comment (yim-lee) [inline Sources/PackageRegistry/SignatureValidation.swift:241]: delete if no longer used

**Review by yim-lee — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
I guess self-hosted are now expected to fail?

**Comment by shahmishal** [comment]
Correct, until we have 5.7 or newer swift.

**Pull request merged**